File: doclens/web_v2/models/pst.py
"""PST 邮件列表 / 附件 API 请求响应模型。"""
from typing import Optional

from pydantic import BaseModel


class PstEmailItem(BaseModel):
    """邮件列表表格的一行。"""
    entry_id: str
    subject: str
    sender: str
    date: str
    folder: str


class PstEmailListResponse(BaseModel):
    """GET /api/pst/emails 响应（日期倒序分页）。"""
    path: str
    total: int
    offset: int
    limit: int
    emails: list[PstEmailItem]


class PstAttachmentInfo(BaseModel):
    """单邮件预览响应中的附件项。

    stored=False（超限/提取失败）时 download_url 为 None，前端只展示文件名。
    """
    name: str
    size: int
    stored: bool
    download_url: Optional[str] = None

File: doclens/web_v2/frontend/src/api/pst.ts
/** GET /api/pst/emails 客户端 + PST 路径判定工具。 */

export interface PstEmailItem {
  entry_id: string;
  subject: string;
  sender: string;
  date: string;
  folder: string;
}

export type PstEmailListResult =
  | {
      ok: true;
      path: string;
      total: number;
      offset: number;
      limit: number;
      emails: PstEmailItem[];
    }
  | { ok: false; notIndexed: boolean; message: string };

/**
 * 调用 GET /api/pst/emails 获取邮件分页列表（日期倒序）。
 * 与 fetchPreview 同一约定：不抛异常，调用方用判别联合处理。
 */
export async function fetchPstEmails(
  path: string,
  offset = 0,
  limit = 50,
): Promise<PstEmailListResult> {
  const params = new URLSearchParams({
    path,
    offset: String(offset),
    limit: String(limit),
  });
  try {
    const res = await fetch(`/api/pst/emails?${params}`);
    if (res.ok) {
      const body = await res.json();
      return {
        ok: true,
        path: body.path,
        total: body.total,
        offset: body.offset,
        limit: body.limit,
        emails: body.emails ?? [],
      };
    }
    const err = await res.json().catch(() => ({ code: "UNKNOWN", detail: "" }));
    return {
      ok: false,
      notIndexed: err.code === "NOT_INDEXED",
      message: err.detail || err.code || `HTTP ${res.status}`,
    };
  } catch (e) {
    return {
      ok: false,
      notIndexed: false,
      message: (e as Error).message || "网络错误",
    };
  }
}

/** PST 物理文件路径（xxx.pst，非派生邮件路径）：预览 = 分页邮件列表。 */
export function isPstFilePath(path: string): boolean {
  return path.toLowerCase().endsWith(".pst") && !path.includes("#");
}

/** PST 派生邮件路径（<pst>#<entry_id>）：预览 = 合成 md + 附件清单。 */
export function isPstEmailPath(path: string): boolean {
  const lower = path.toLowerCase();
  return lower.includes("#") && lower.split("#")[0].endsWith(".pst");
}
